crates/wysiwyg/src/node_joiner.rs

```diff
@@ -48,34 +46,285 @@ where
+                ) = (prev_node.kind(), next_node.kind())
+                {
+                    // Found a matching sibling node with the same format
+                    if prev_format == next_format {
+                        // Move the contents from the current node to the previous one
+                        let new_index =
+                            Self::move_children(dom, &cur_handle, &prev_handle);
+                        // Next iteration
+                        let mut cur_path = handle.raw().clone();
+                        let prev_path = prev_handle.raw();
+                        cur_path[level - 1] = prev_path[level - 1];
+                        cur_path[level] = new_index;
+                        let new_handle = DomHandle::from_raw(cur_path);
+
+                        self.join_format_nodes_at_level(
+                            dom,
+                            new_handle,
+                            level + 1,
+                        );
+                    }
+                } else {
+                    // If both nodes weren't formatting nodes, try at the next level
+                    self.join_format_nodes_at_level(dom, handle, level + 1);
+                }
+            }
+        } else {
+            // If there's no previous node, try at the next level
+            self.join_format_nodes_at_level(dom, handle, level + 1);
+        }
+    }
+
+    fn do_join(
+        dom: &mut Dom<S>,
+        ancestors_start: &Vec<DomHandle>,
+        ancestors_next: &Vec<DomHandle>,
+    ) {
+        let mut i = 0;
+        let mut j = 0;
+        while i < ancestors_start.len() && j < ancestors_next.len() {
+            let start_handle = ancestors_start.get(i).unwrap();
+            let next_handle = ancestors_next.get(j).unwrap();
+
+            // If both lists contain this ancestor handle, continue to the next comparison.
+            if start_handle == next_handle {
+                i += 1;
+                j += 1;
+                continue;
+            }
+
+            let start_i = dom.lookup_node(start_handle.clone());
+            let next_i = dom.lookup_node(next_handle.clone());
+
+            match (start_i, next_i) {
+                (DomNode::Container(start_i), DomNode::Container(next_i)) => {
+                    // TODO: check if this is_structure_node verification is needed.
+                    if start_i.is_structure_node()
+                        && next_i.is_structure_node()
+                        && start_i.name() == next_i.name()
+                    {
+                        // Both containers with the same tag.
+                        // Move children from next to start node, remove next node.
+                        let new_index_in_parent =
+                            Self::move_children(dom, next_handle, start_handle);
+                        // This alters ancestors, so we need to re-calculate them and start again.
+                        // TODO: maybe move this to another fn?
```

Comment:
Would be good to shorten this function.

Reply:
I tried extracting the re-calculation of ancestors to another function. I'm thinking about extracting text node merging too, I'll see if it's easy to do.